ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java

```diff
@@ -252,15 +236,49 @@ private void onChunkOperationComplete(GetChunk chunk) {
             routerMetrics.onGetBlobError(e, options, isEncrypted);
           }
         }
-        NonBlockingRouter.completeOperation(null, getOperationCallback, operationResult, e);
+        // We don't complete operation in case of segmented blob, as we don't have all the header information until we fetch the segmented chunk.
+        if (e != null || !options.getBlobOptions.hasBlobSegmentIdx()) {
+          NonBlockingRouter.completeOperation(null, getOperationCallback, operationResult, e);
+        }
       }
+    } else if (options.getBlobOptions.hasBlobSegmentIdx()) {
+      // We have fetched the segmented chunk, and there cannot be any more chunks to fetch now.
+      fixBlobSizeIfRequired(blobInfo.getBlobProperties().getBlobSize(),
+          firstChunk.getChunkMetadataList().get(0).getSize());
```

Comment:
This seems to always set the size to the size of the first chunk regardless of which chunk is being fetched. Also, why can't this be done in the `chunk == firstChunk` block above? It looks like we have all of the necessary pieces to complete the operation at that time (`BlobInfo` for the properties of the composite blob, `ChunkMetadata` list to get the datachunk size, segment index from options)

Reply:
You are right about being able to do it in the firstChunk block itself. fixed it!